### ISSUE TITLE
pause, resume for notification in ios 

### ISSUE THREAD

**Issue by Shahmirzali-Huseynov** [781flyingdutchman/background_downloader#75]
the app show notification only when the file download is completed. are there some functionalities in the notification for  pause, resume or it send just when the file download completed? 

note: I tested your example folder sample

<img width="469" alt="Screenshot 2023-06-16 at 23 50 10" src="https://github.com/781flyingdutchman/background_downloader/assets/49102327/eb281dba-5311-4f3d-9f55-53e69a4c7f48">

**Comment by 781flyingdutchman** [comment]
As configured in the example app, it also shows a 'running' notification, though on iOS, per their guidelines, it's a static notification and only shows in the notification center (so, unlike with Android, it doesn't show a progress bar etc, and it doesn't pop into view when the app is in the foreground, which is why you don't see it when you start in the foreground and then go back to the home screen - but it is there, in the notification center).

If the task can pause (which in the example app is only the case for the download associated with the main download button - not for the 'Load a bunch' files, which you show in your screenshot, as those do not have `allowPause` set to `true` ), then the 'running' notification can be expanded to show a pause button, and if it is paused (also via the in-app button or progress bar) it will show the 'paused' notification from where you can resume. See lines 60-71 for the configuration used.

All this said, I am having a hard time myself getting the pause button to appear right now in the iOS simulator so I may take another look to see if something broke.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman  I tested a single download (main download button). I used a 90MB mp3 file. after pressing the download button I went to home screen or other apps. and waited for notification 🙂, but the notification arrived just when the file download was completed.

**Comment by 781flyingdutchman** [comment]
Apple does notifications differently. See the [Notifications](https://pub.dev/packages/background_downloader#notifications) section of the README, and the [Apple guidelines](https://developer.apple.com/design/human-interface-guidelines/notifications) listed there, further complicated by how the iOS simulators work. So, to summarize and record for posterity:

- Notifications are not shown when the app is in the foreground, only when it is not. When the app is in the foreground, notifications are available in the notification center, which you open by swiping from the top of the screen.  So, when you start the download by tapping the Download button, you are in the foreground and the 'running' notification is therefore only visible in the notification center. If you go to your home screen and do nothing else, then the next thing that will show is the 'complete' notification, because that is triggered when the app is not in the foreground. To see the 'running' notification, don't wait for 'complete' and swipe down from the top to see the notification center. The 'running' notification is there.
- The 'running' notification does not update. Unlike Android, the 'running' notification is only triggered once, at the start of the download, and not updated. It therefore does not show a progress bar, and it does not substitute {progress} with a live progress indication. This is per Apple guidelines
- Now that you see the 'running' notification you wonder why you don't see 'pause' and 'cancel'. In iOS, action buttons only show when you expose them. On devices with 3D touch you use that on the notification itself (a bit like a long press on Android, where you push a bit harder), and on devices without 3D touch you need to swipe the notification slightly sideways (left, I think), to expose them. The buttons will then show below the notification. Not ideal, but that's the UI
- On a simulator it is even harder: if you run on a simulator that supports 3D touch, yet your Mac does not, then there is no way for you to expose those buttons. And if your Mac does support 3D touch, make sure the Simulator also has that enabled. The reason I mentioned earlier that I'm having a hard time getting the button to appear in the simulator is exactly that: it works fine on a real device, I can't get it to show in the simulator.

In summary, I am confident the notifications work as intended on iOS, and I suspect that you - like me - do not find the notification UI on iOS particularly easy. But I'm following the guidelines, so don't want to change anything.

Hope this helps.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman When I tested in a real device iPhone, it just arrived Completed notification. I simulate all situations, NotificationCenter and others but nothing happened. Only completed notification arrived when the download was finished.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman If you have a real device iPhone can you same make a video example about notification for ReadMe ?

**Comment by 781flyingdutchman** [comment]
I'm sorry but that's not something I feel I need to do. The example app works exactly as I just described it above. If it's going to fast, then perhaps modify the url to a bigger file so it doesn't complete before you can look at the 'running' notification.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman I tested 90MB file as I said before. just arrived completed notification. I tested even 2 real devices.

**Comment by 781flyingdutchman** [comment]
Try inserting `os_log("content.categoryIdentifier %@", log: log, type: .error, content.categoryIdentifier)` after line 94 in `Notifications.swift`. The console should now log `content.categoryIdentifier running_with_pause` or `content.categoryIdentifier running_without_pause` when the running notification gets created (you can filter on Category = Downloader to remove clutter). If it does, then really the notification gets created. If it doesn't, and you use the standard example app and the regular download button, then I don't know why not.

**Comment by 781flyingdutchman** [comment]
Also, try the standard example app (latest version, Dart 3) and change the URL in line 219 from 5MB-test to 57MB-test. It gives you more time to check the notifications, and it really does work on my device, running iOS 16.5. Remember to use the topmost 'Download' button, not any of the others.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman I will test and reply here. but the logs will be in Xcode or can I see logs in vs code debug console? because I will test in the release mod real device

**Comment by 781flyingdutchman** [comment]
I don't think the logs will show in release mode. I look at iOS logs in the Console app on Mac, and am not sure they'll show up in your regular Flutter logs. But try the stock app with the URL change I suggested too - if that works then the issue is elsewhere

**Comment by 781flyingdutchman** [comment]
I just realized another thing: you **must** use a **unique** taskId for every download task. Even between app restarts you cannot use the same taskId - if you do, the progress notification will NOT be triggered. 

Also, what device and iOS version are you using?

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman I just testing your example. in your example, taskId isn't unique ? 

my device : iPhone 12 pro max , ios version 16.5

**Comment by 781flyingdutchman** [comment]
In the example it's unique, as it's omitted when the task is created and therefore becomes a random number. I was just thinking that if it's not unique in how you use it, that may explain it.

**Comment by Shahmirzali-Huseynov** [comment]
@

> I don't think the logs will show in release mode. I look at iOS logs in the Console app on Mac, and am not sure they'll show up in your regular Flutter logs. But try the stock app with the URL change I suggested too - if that works then the issue is elsewhere

You are mac console app but what are you using for running app ? Xcode or vs code ?

**Comment by 781flyingdutchman** [comment]
I'm using Android Studio, but of course it uses Xcode to compile for iOS

**Comment by 781flyingdutchman** [comment]
Also, check in your iPhone Settings | Notifications and the specific notification settings for your app that you have enabled alerts are showing in the notification center (and not only banners. Perhaps your notification center settings are different

**Comment by Shahmirzali-Huseynov** [comment]
I inserted os_log("content.categoryIdentifier %@", log: log, type: .error, content.categoryIdentifier) after line 94 in Notifications.swift.

<img width="870" alt="Screenshot 2023-06-18 at 19 07 06" src="https://github.com/781flyingdutchman/background_downloader/assets/49102327/1248aa53-d295-4eb9-aafe-87963744b747">

Console Mac app logs. @781flyingdutchman are logs what you excepted?

<img width="824" alt="Screenshot 2023-06-18 at 19 07 26" src="https://github.com/781flyingdutchman/background_downloader/assets/49102327/d17f65d6-d32e-4111-8469-658ff8dd49af">

My phone's and app's notification settings :

![IMG_0146](https://github.com/781flyingdutchman/background_downloader/assets/49102327/b2d8781d-1239-4a76-b40e-b037bc70f4c4)

![IMG_0147](https://github.com/781flyingdutchman/background_downloader/assets/49102327/350cd641-3765-4221-a72e-b6ac023ab43b)

Notification arrived finally but just when I open notification center, even when I background. most of the time users do not open NotificationCenter everytime. We probably have to use local notification package.

<img width="556" alt="Screenshot 2023-06-18 at 19 10 21" src="https://github.com/781flyingdutchman/background_downloader/assets/49102327/93336778-eed1-4fe3-9c52-3b2d47488345">

<img width="555" alt="Screenshot 2023-06-18 at 19 10 32" src="https://github.com/781flyingdutchman/background_downloader/assets/49102327/90fb74af-606e-4340-9d63-87646f9e2a5d">

and also when I pause from notification. paused notification arrived and gone just for 1 second 😅.

**Comment by 781flyingdutchman** [comment]
Hi, so this indeed looks like it works as intended. The Apple Guideline is that you do not show notifications when the app is in the foreground (because you should notify the user through your app's UI - for example the download progress widget), and therefore those notifications move to the notification center. You can choose a different approach (as could I) but just know that that behavior is not what an iPhone user expects: they already know the download is in progress and don't need to see another notification saying so - they just want to see when it completes (and they can). If a download starts while the app is in the _background_ (e.g. because you have a long queue of downloads) then _those_ 'running' notifications will show up as notifications to the user, because it's news to them. Again, this is as per Apple Guidelines - if you start spamming users with notifications you may have an issue getting this pas Apple's app approvals :)

As for the pause notification disappearing almost immediately: I noticed that too, and have fixed that - will be part of the next release as a bug fix.

**Comment by Shahmirzali-Huseynov** [comment]
@781flyingdutchman Okay, thanks for your reply, effort, time, and this package.
